In [1]:
import pandas as pd

In [2]:
import numpy as np

# doc_list

In [3]:
doc_list = pd.read_csv("doctors_list_t8.csv")

In [4]:
pd.isnull(doc_list).sum()

Id             0
initial        0
href           0
name           0
specialties    0
address        0
zip            0
dtype: int64

In [5]:
doc_list.head()

Id initial                         href               name  \
0  3886560       a  /doctors/Dr_Enenge_ABodjedi    Enenge Abodjedi   
1  3886561       a     /doctors/Dr_Brady_AHearn    Brady J A'Hearn   
2  3886562       a    /doctors/Dr_Yousif_Arahim     Yousif A-Rahim   
3  3886563       a    /doctors/Dr_Noor_Al_Khori      Noor Al Khori   
4  3886564       a  /doctors/Dr_Salam_Al-Hafidh  Salam N Al-Hafidh   

                                        specialties                   address  \
0                                  ['Psychiatrist']        Stamford, CT 06902   
1                                  ['Pediatrician']       Iowa City, IA 52242   
2  ['Gastroenterologist (digestive)', ' Internist']            Aiea, HI 96701   
3                                       ['Student']    Grand Rapids, MI 49503   
4                                  ['Pediatrician']  Richland Hills, TX 76180   

     zip  
0  06902  
1  52242  
2  96701  
3  49503  
4  76180

In [6]:
doc_list.tail()

Id initial                        href            name  \
1294498  5181716       i   /dentists/Dr_Micheal_Izzo  Michael A Izzo   
1294499  5181717       i     /doctors/Dr_Philip_Izzo   Philip M Izzo   
1294500  5181718       i    /dentists/Dr_Steven_Izzo   Steven R Izzo   
1294501  5181719       i  /dentists/Dr_Steven_Izzo_2     Steven Izzo   
1294502  5181720       i    /doctors/Dr_Timothy_Izzo  Timothy J Izzo   

                                             specialties  \
1294498                                      ['Dentist']   
1294499                                 ['Psychiatrist']   
1294500                                      ['Dentist']   
1294501                                      ['Dentist']   
1294502  ['Family Practitioner', ' Primary Care Doctor']   

                        address    zip  
1294498  Lincoln Park, MI 48146  48146  
1294499         Media, PA 19063  19063  
1294500     Ridgewood, NY 11385  11385  
1294501      Brooklyn, NY 11228  11228  
1294502   Grand Ledge, MI 48837  48837

In [7]:
np.shape(doc_list) # see the number of rows and columns

(1294503, 7)

- We have 1294503 doctors in the data set

In [8]:
r = lambda x: x.lstrip("/") 

In [9]:
doc_list.href = doc_list.href.apply(r) # split the href into title and the prof_name

In [10]:
doc_list['title'], doc_list['prof_name'] = doc_list['href'].str.split('/',1).str

In [11]:
doc_list = doc_list.drop('href',1)

In [12]:
doc_list.tail() 

Id initial            name  \
1294498  5181716       i  Michael A Izzo   
1294499  5181717       i   Philip M Izzo   
1294500  5181718       i   Steven R Izzo   
1294501  5181719       i     Steven Izzo   
1294502  5181720       i  Timothy J Izzo   

                                             specialties  \
1294498                                      ['Dentist']   
1294499                                 ['Psychiatrist']   
1294500                                      ['Dentist']   
1294501                                      ['Dentist']   
1294502  ['Family Practitioner', ' Primary Care Doctor']   

                        address    zip     title         prof_name  
1294498  Lincoln Park, MI 48146  48146  dentists   Dr_Micheal_Izzo  
1294499         Media, PA 19063  19063   doctors    Dr_Philip_Izzo  
1294500     Ridgewood, NY 11385  11385  dentists    Dr_Steven_Izzo  
1294501      Brooklyn, NY 11228  11228  dentists  Dr_Steven_Izzo_2  
1294502   Grand Ledge, MI 48837  48837   doctors   Dr_Timothy_Izzo

In [54]:
doc_list_dupli = doc_list[doc_list.duplicated(['prof_name'], keep=False)]

In [55]:
doc_list_dupli_name = set(doc_list_dupli.prof_name)

In [57]:
len(doc_list_dupli_name)

0

In [13]:
doc_list.title.value_counts()

doctors     1095990
dentists     198513
Name: title, dtype: int64

There are 1095990 doctors and 198513 dentists

In [14]:
import re

In [15]:
rs = lambda x: re.sub('[^a-zA-Z\']+', ' ', x)

In [16]:
doc_list.specialties = doc_list.specialties.apply(rs) # clean the specialties columns

In [17]:
doc_list.head()

Id initial               name  \
0  3886560       a    Enenge Abodjedi   
1  3886561       a    Brady J A'Hearn   
2  3886562       a     Yousif A-Rahim   
3  3886563       a      Noor Al Khori   
4  3886564       a  Salam N Al-Hafidh   

                                      specialties                   address  \
0                                 'Psychiatrist'         Stamford, CT 06902   
1                                 'Pediatrician'        Iowa City, IA 52242   
2   'Gastroenterologist digestive ' ' Internist'             Aiea, HI 96701   
3                                      'Student'     Grand Rapids, MI 49503   
4                                 'Pediatrician'   Richland Hills, TX 76180   

     zip    title           prof_name  
0  06902  doctors  Dr_Enenge_ABodjedi  
1  52242  doctors     Dr_Brady_AHearn  
2  96701  doctors    Dr_Yousif_Arahim  
3  49503  doctors    Dr_Noor_Al_Khori  
4  76180  doctors  Dr_Salam_Al-Hafidh

In [18]:
np.shape(doc_list)

(1294503, 8)

In [19]:
words = []
for i in doc_list.specialties:
    words += [i]

In [20]:
words_clean = re.sub('[^a-zA-Z]+', '/', str(words))

In [21]:
#count the specialties
specialist = {}
wordsplit = words_clean.split('/')
for word in wordsplit:
    if word in specialist :
        specialist[word]=specialist[word]+1
    else:
        specialist[word] = 1
     

In [22]:
sorted(specialist.items(), key=lambda x: x[1],reverse=True)

[('Specialist', 229772),
 ('Internist', 222313),
 ('Dentist', 198510),
 ('Care', 151698),
 ('Practitioner', 147143),
 ('Primary', 139732),
 ('Doctor', 139732),
 ('Family', 131692),
 ('Surgical', 100403),
 ('Pediatrician', 92566),
 ('Psychologist', 87175),
 ('Psychiatrist', 64219),
 ('Student', 56887),
 ('Obstetrician', 52807),
 ('Gynecologist', 52807),
 ('OB', 52807),
 ('GYN', 52807),
 ('Anesthesiologist', 52793),
 ('pain', 52793),
 ('control', 52793),
 ('Surgeon', 49230),
 ('Radiologist', 48060),
 ('Emergency', 46557),
 ('Physician', 46557),
 ('Orthopaedic', 31876),
 ('Cardiologist', 31593),
 ('heart', 31593),
 ('Oncology', 26454),
 ('Medicine', 26209),
 ('Pediatric', 23938),
 ('Pathologist', 23725),
 ('Ophthalmologist', 23133),
 ('eyes', 23133),
 ('cancer', 22076),
 ('Neurologist', 20461),
 ('brain', 20461),
 ('nervous', 20461),
 ('system', 20460),
 ('Gastroenterologist', 16900),
 ('digestive', 16900),
 ('General', 15451),
 ('Dermatologist', 14867),
 ('skin', 14867),
 ('Pulmonologist

the TOP 10 	specialties are:
 - ('Internist', 222313),
 - ('Dentist', 198510),
 - ('PrimaryCareDoctor', 139732),
 - ('FamilyPractitioner', 131692),
 - ('SurgicalSpecialist', 100403),
 - ('Pediatrician', 92566),
 - ('Psychologist', 87175),
 - ('Psychiatrist', 57374),
 - ('Student', 56887),
 - ('ObstetricianGynecologistOBGYN', 52807)

In [23]:
specialist_table = pd.DataFrame.from_dict(specialist, orient='index')

# doc_panel

In [24]:
doc_panel = pd.read_csv("doctors_panel_t8.csv")

In [25]:
doc_panel.head()

Id                                href  years_of_exp  overall_rating  \
0  9206066         /dentists/Dr_Aamir_M_Sheikh          10.0             0.0   
1  9206067           /dentists/Dr_Aamir_Shah_2           NaN             4.5   
2  9206068            /dentists/Dr_Aalie_Sagar           NaN             3.5   
3  9206069           /dentists/Dr_Aan_Litvinov           NaN             1.0   
4  9206070  /dentists/Dr_Aarika_Anderson_Elter           NaN             5.0   

   n_ratings  n_reviews avg_wating  Ease_of_Appointment  Promptness  \
0          0          0        NaN                  NaN         NaN   
1          3          0        NaN                  NaN         NaN   
2          3          0        NaN                  NaN         NaN   
3          1          0        NaN                  NaN         NaN   
4          1          1        NaN                  3.5         NaN   

   Courteous_Staff         ...           Caring_Manner  Pain_Minimized  \
0              NaN         ...                     NaN             NaN   
1              NaN         ...                     NaN             NaN   
2              NaN         ...                     NaN             NaN   
3              NaN         ...                     NaN             NaN   
4              NaN         ...                     5.0             5.0   

   Satisfaction  n_5  n_4  n_3  n_2  n_1  \
0           NaN  NaN  NaN  NaN  NaN  NaN   
1           NaN  NaN  NaN  NaN  NaN  NaN   
2           NaN  NaN  NaN  NaN  NaN  NaN   
3           NaN  NaN  NaN  NaN  NaN  NaN   
4           NaN  1.0  0.0  0.0  0.0  0.0   

                                     address           crawl_date  
0      9105 All Saints Rd, Laurel, MD, 20723  2016-09-06 20:53:49  
1        7872 Walker St, La Palma, CA, 90623  2016-09-06 20:53:51  
2  10842 Potranco Rd, San Antonio, TX, 78251  2016-09-06 20:53:54  
3       1740 Culver Rd, Rochester, NY, 14609  2016-09-06 20:53:55  
4          1708 E 44th St, Tacoma, WA, 98404  2016-09-06 20:54:01  

[5 rows x 27 columns]

In [27]:
np.shape(doc_panel)

(1293682, 27)

In [30]:
doc_panel.href = doc_panel.href.apply(r) # split the href into title and the prof_name
doc_panel['title'], doc_panel['prof_name'] = doc_panel['href'].str.split('/',1).str
doc_panel = doc_panel.drop('href',1)

In [31]:
doc_panel.tail()

Id  years_of_exp  overall_rating  n_ratings  n_reviews  \
1293677  10499745          25.0             4.5         21         11   
1293678  10499746          11.0             0.0          0          0   
1293679  10499747          37.0             3.0          2          1   
1293680  10499748          26.0             4.5         18          1   
1293681  10499749          49.0             5.0          2          0   

         avg_wating  Ease_of_Appointment  Promptness  Courteous_Staff  \
1293677  33 minutes                  5.0         4.0              4.5   
1293678         NaN                  NaN         NaN              NaN   
1293679  10 minutes                  5.0         3.5              5.0   
1293680  17 minutes                  4.0         4.5              4.5   
1293681         NaN                  NaN         NaN              NaN   

         Accurate_Diagnosis        ...          Satisfaction   n_5  n_4  n_3  \
1293677                 5.0        ...                   NaN  16.0  2.0  1.0   
1293678                 NaN        ...                   NaN   NaN  NaN  NaN   
1293679                 5.0        ...                   NaN   1.0  0.0  0.0   
1293680                 4.5        ...                   NaN  14.0  1.0  0.0   
1293681                 NaN        ...                   NaN   NaN  NaN  NaN   

         n_2  n_1                                 address  \
1293677  2.0  0.0  10293 N 92nd St, Scottsdale, AZ, 85258   
1293678  NaN  NaN   60 Hospital Rd, Leominster, MA, 01453   
1293679  0.0  1.0     200 1st St Sw, Rochester, MN, 55905   
1293680  0.0  3.0    1155 Jadwin Ave, Richland, WA, 99352   
1293681  NaN  NaN      3909 Woodley Rd, Toledo, OH, 43606   

                  crawl_date    title          prof_name  
1293677  2016-09-16 20:03:43  doctors       Dr_Ramy_Doss  
1293678  2016-09-16 20:03:45  doctors    Dr_Ramzi_Khazen  
1293679  2016-09-16 20:03:45  doctors   Dr_Randall_Edson  
1293680  2016-09-16 20:03:47  doctors  Dr_Randall_Farrar  
1293681  2016-09-16 20:03:48  doctors  Dr_Randall_Finken  

[5 rows x 28 columns]

In [32]:
doc_panel.columns

Index(['Id', 'years_of_exp', 'overall_rating', 'n_ratings', 'n_reviews',
       'avg_wating', 'Ease_of_Appointment', 'Promptness', 'Courteous_Staff',
       'Accurate_Diagnosis', 'Bedside_Manner', 'Spends_Time_with_Me',
       'Follows_Up_After_Visit', 'Comfortable_Office', 'Cleanliness',
       'Advanced_Technology', 'Caring_Manner', 'Pain_Minimized',
       'Satisfaction', 'n_5', 'n_4', 'n_3', 'n_2', 'n_1', 'address',
       'crawl_date', 'title', 'prof_name'],
      dtype='object')

In [36]:
doc_panel.isnull().values.any()

True

In [37]:
doc_panel.isnull().sum()

Id                              0
years_of_exp               432050
overall_rating                  0
n_ratings                       0
n_reviews                       0
avg_wating                1000733
Ease_of_Appointment        859795
Promptness                 895034
Courteous_Staff            893877
Accurate_Diagnosis         891816
Bedside_Manner             891785
Spends_Time_with_Me        891773
Follows_Up_After_Visit     897890
Comfortable_Office        1259010
Cleanliness               1259017
Advanced_Technology       1259167
Caring_Manner             1258985
Pain_Minimized            1259259
Satisfaction              1259697
n_5                        844215
n_4                        844215
n_3                        844215
n_2                        844215
n_1                        844216
address                         0
crawl_date                      0
title                           0
prof_name                       0
dtype: int64

In [49]:
doc_panel_dupli = doc_panel[doc_panel.duplicated(['prof_name'], keep=False)]

Id  years_of_exp  overall_rating  n_ratings  n_reviews  \
19216     9225282          22.0             5.0        324        323   
19311     9225377          22.0             5.0        324        323   
26428     9232494          14.0             5.0        374        373   
26518     9232584          14.0             5.0        374        373   
50853     9256919          20.0             4.5        456        450   
50944     9257010          20.0             4.5        456        450   
101851    9307917           NaN             5.0        425        423   
101920    9307986           NaN             5.0        425        423   
119901    9325967           NaN             5.0        588        586   
120010    9326076           NaN             5.0        588        586   
139972    9346038          27.0             5.0        488        488   
140066    9346132          27.0             5.0        488        488   
140122    9346188          27.0             5.0        488        488   
140784    9346850          19.0             5.0        181        181   
140876    9346942          19.0             5.0        181        181   
147191    9353257           NaN             5.0        166        164   
147269    9353335           NaN             5.0        166        164   
156520    9362586           NaN             5.0        167        164   
156592    9362658           NaN             5.0        167        164   
160591    9366657           NaN             5.0        297        296   
160606    9366672           NaN             5.0        297        296   
170046    9376112          13.0             5.0        425        423   
170116    9376182          13.0             5.0        425        423   
170130    9376196          13.0             5.0        425        423   
170746    9376812           NaN             5.0        350        347   
170880    9376946           NaN             5.0        350        347   
171874    9377940           NaN             5.0        159        158   
171954    9378020           NaN             5.0        159        158   
175822    9381888           NaN             5.0        154        152   
175881    9381947           NaN             5.0        154        152   
...           ...           ...             ...        ...        ...   
1293239  10499306          24.0             5.0        624        526   
1293240  10499307          24.0             5.0        624        526   
1293241  10499308          24.0             5.0        624        526   
1293242  10499309          24.0             5.0        624        526   
1293243  10499310          24.0             5.0        624        526   
1293244  10499311          24.0             5.0        624        526   
1293245  10499312          24.0             5.0        624        526   
1293246  10499313          24.0             5.0        624        526   
1293247  10499314          24.0             5.0        624        526   
1293248  10499315          24.0             5.0        624        526   
1293249  10499316          24.0             5.0        624        526   
1293250  10499317          24.0             5.0        624        526   
1293251  10499318          24.0             5.0        624        526   
1293252  10499319          24.0             5.0        624        526   
1293253  10499320          24.0             5.0        624        526   
1293254  10499321          24.0             5.0        624        526   
1293255  10499322          24.0             5.0        624        526   
1293256  10499323          24.0             5.0        624        526   
1293257  10499324          24.0             5.0        624        526   
1293258  10499325          24.0             5.0        624        526   
1293259  10499326          24.0             5.0        624        526   
1293260  10499327          24.0             5.0        624        526   
1293261  10499328          24.0             5.0        624        526   


In [50]:
doc_panel_dupli_name = set(doc_panel_dupli.prof_name)

In [52]:
len(doc_panel_dupli_name)

174

In [53]:
doc_panel_dupli.prof_name.value_counts()

Dr_Allan_Sawyer           84
Dr_Francis_Clark           6
Dr_Shane_Matt              3
Dr_Ralph_Pearlman          3
Dr_Scott_M_Schlesinger     3
Dr_Patricia_Cleavinger     3
Dr_Terry_Test              3
Dr_William_Romero          3
Dr_Joseph_B_Wilson         3
Dr_Sharad_Rajpal           2
Dr_Siamak_Ourian           2
Dr_William_Roddy           2
Dr_Boris_Abayev            2
Dr_Gregg_Friedman          2
Dr_Bret_Lewis              2
Dr_Warren_Ravner           2
Dr_Monica_Mehta            2
Dr_Kevin_Snodgrass         2
Dr_Randall_Dryer           2
Dr_Linda_C_Huang           2
Dr_Farzad_Massoudi         2
Dr_Barry_J_Cohen           2
Dr_Wilberto_Cortes         2
Dr_Amer_Qazi               2
Dr_Robert_Rothfield        2
Dr_Ara_Deukmedjian         2
Dr_Adam_Angeles            2
Dr_Jeffrey_Glaser          2
Dr_Sidney_Wu               2
Dr_Daniel_Feingold         2
                          ..
Dr_David_Zelouf            2
Dr_Jarrod_Kaufman          2
Dr_Michael_J_Duffy         2
Dr_Marvin_Spon

# campare list and panel table

In [136]:
namelist = list(doc_list.prof_name)
len(namelist)

1294503

In [79]:
checklist = list(doc_panel.prof_name)
len(checklist)

1293682

In [94]:
len(set(checklist))

1293415

In [129]:
doc_inTwoTable = list(set(checklist).intersection(namelist))

In [130]:
len(doc_inTwoTable)

1292734

In [131]:
miss_inPanel = set(checklist)-set(doc_inTwoTable)

In [133]:
len(miss_inPanel)

681

In [134]:
miss_inList = set(namelist) - set(doc_inTwoTable)

In [135]:
len(miss_inList)

1769

# reviews_panel

In [29]:
rew_panel = pd.read_csv("reviews_panel_t8.csv")

In [32]:
np.shape(rew_panel)

(1796204, 23)

In [35]:
3886560 in rew_panel.review_id

False

In [8]:
rew_panel.head()

Id  review_id                                href  overall_rating  \
0  12949966   23507540  /dentists/Dr_Aarika_Anderson_Elter               5   
1  12949967   27184431            /dentists/Dr_Aamir_Wahab               5   
2  12949968   26307282           /dentists/Dr_Aanal_Parikh               1   
3  12949969   28904504          /dentists/Dr_Aaron_Aguilar               5   
4  12949970   28380953         /dentists/Dr_Aaron_D_Larsen               4   

   Ease_of_Appointment  Promptness  Courteous_Staff  Accurate_Diagnosis  \
0                  4.0         NaN              NaN                 NaN   
1                  5.0         NaN              NaN                 NaN   
2                  1.0         NaN              NaN                 NaN   
3                  5.0         NaN              NaN                 NaN   
4                  3.0         NaN              NaN                 NaN   

   Bedside_Manner  Spends_Time_with_Me         ...           \
0             NaN                  NaN         ...            
1             NaN                  NaN         ...            
2             NaN                  NaN         ...            
3             NaN                  NaN         ...            
4             NaN                  NaN         ...            

   Advanced_Technology  Caring_Manner  Pain_Minimized  Satisfaction  \
0                  NaN            5.0             5.0           NaN   
1                  5.0            5.0             5.0           5.0   
2                  1.0            1.0             1.0           1.0   
3                  5.0            5.0             5.0           5.0   
4                  4.0            4.0             4.0           5.0   

         date      reviewer                                 title  \
0  2012-11-16           NaN                        Great results    
1  2015-03-15       Brad s.                               implant   
2  2014-10-12  Bill johnson  Warning ..would not see this dentist   
3  2015-06-11           NaN                                   NaN   
4  2015-03-31           NaN             Dr larsen great others no   

                                             content helpful_vote  \
0  Dr. Anderson explained in detail my options fo...          NaN   
1  Had an implant done and it was painless. I cou...          NaN   
2  Do not go to this dentist office they scam peo...          NaN   
3  Very thoughtful Dr.  Communicates/bedside mann...          NaN   
4  Would continue to see Dr.  Larsen however his ...          NaN   

            crawl_date  
0  2016-09-06 20:54:01  
1  2016-09-06 20:54:10  
2  2016-09-06 20:54:13  
3  2016-09-06 20:54:14  
4  2016-09-06 20:54:18  

[5 rows x 23 columns]

In [12]:
rew_panel.tail()

Id  review_id                        href  overall_rating  \
1796199  14819364   19546944       /doctors/Dr_Ramy_Doss               5   
1796200  14819365   18196930       /doctors/Dr_Ramy_Doss               5   
1796201  14819366   18233638       /doctors/Dr_Ramy_Doss               5   
1796202  14819367   19725098   /doctors/Dr_Randall_Edson               5   
1796203  14819368   31104374  /doctors/Dr_Randall_Farrar               5   

         Ease_of_Appointment  Promptness  Courteous_Staff  Accurate_Diagnosis  \
1796199                  5.0         4.0              5.0                 5.0   
1796200                  5.0         4.0              5.0                 5.0   
1796201                  5.0         4.0              5.0                 5.0   
1796202                  5.0         4.0              5.0                 5.0   
1796203                  5.0         5.0              5.0                 5.0   

         Bedside_Manner  Spends_Time_with_Me         ...           \
1796199             5.0                  5.0         ...            
1796200             5.0                  5.0         ...            
1796201             5.0                  5.0         ...            
1796202             5.0                  5.0         ...            
1796203             5.0                  NaN         ...            

         Advanced_Technology  Caring_Manner  Pain_Minimized  Satisfaction  \
1796199                  NaN            NaN             NaN           NaN   
1796200                  NaN            NaN             NaN           NaN   
1796201                  NaN            NaN             NaN           NaN   
1796202                  NaN            NaN             NaN           NaN   
1796203                  NaN            NaN             NaN           NaN   

               date        reviewer                     title  \
1796199  2011-07-13             NaN                       NaN   
1796200  2008-06-23         Cheri R  Wonderful, caring Doctor   
1796201  2008-04-30  THERESA DALTON                       NaN   
1796202  2011-11-21             NaN                       NaN   
1796203  2015-12-19             NaN                       NaN   

                                                   content helpful_vote  \
1796199  Love  Dr. Army  Doss.  He saved my life- a hea...          NaN   
1796200  I met Dr. Doss during a very stressful time an...          NaN   
1796201  IF IT WEREN'T FOR DR. DOSS' ACCURATE DIAGNOSIS...          NaN   
1796202  dr. edson is as good as it gets for being pers...          1.0   
1796203                           Great Doc! Really cares.          NaN   

                  crawl_date  
1796199  2016-09-16 20:03:44  
1796200  2016-09-16 20:03:44  
1796201  2016-09-16 20:03:44  
1796202  2016-09-16 20:03:45  
1796203  2016-09-16 20:03:47  

[5 rows x 23 columns]